### PR TITLE
Set up email notifications for `user_measures`

### LINKED ISSUE #50: User assignment: set up email notifications for users being assigned to measures (all types)
We need to send email notifications to users when they have been assigned to a measure (of any measuretype)

Specifically, this can be triggered by two changes to the database

- [ ] when a relationship between a measure and a user is established in the "user_measures" table, if the measure is already published (attribute: `draft = false`)
- [ ] when a measure is being published (attribute: `draft` changing from `true` to `false`), if a measure-user relationship has already been established in the "user_measures" table

In addition the following conditions need to be met:

- [ ] the assigned user is not the same user updating the database as specified above
- [ ] activity is not archived `is_archive != true` 
- [ ] activity notifications are enabled `notifications = true` (see #47) 
- [ ] to be confirmed: activity does not have a "measure_categories" relationship with `category_id = 24` ("completed", should be configurable)

#### Email: 

- [ ] Emails should contain link to the measure (will require storing the client-side path in the server configuration)

### PR DESCRIPTION
**Pull request by parndt**
Resolves #50

We need to send email notifications to users when they have been assigned to a measure (of any measuretype)

Specifically, this can be triggered by two changes to the database

- [x] when a relationship between a measure and a user is established in the "user_measures" table, if the measure is already published (attribute: `draft = false`)
- [x] when a measure is being published (attribute: `draft` changing from `true` to `false`), if a measure-user relationship has already been established in the "user_measures" table

In addition the following conditions need to be met:

- [x] the assigned user is not the same user updating the database as specified above
- [x] activity is not archived `is_archive != true` 
- [x] activity notifications are enabled `notifications = true` (see #47) 
- [ ] to be confirmed: activity does not have a "measure_categories" relationship with `category_id = 24` ("completed", should be configurable)

#### Email: 

- [x] Emails should contain link to the measure (will require storing the client-side path in the server configuration)

### CODE REVIEW

File: app/views/user_measure_mailer/created.en.html.erb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<h1>Measure Assigned</h1>
+
+<p>
+  Hi <%= @name %>,
+</p>
+
+<p>
+  You have a new measure assigned to you <%= @title %>.
```
Comment (tmfrnz) [inline app/views/user_measure_mailer/created.en.html.erb:8]: This should also include a link to the measure which is
https://plastic-advocacy.wwf.no/action/[id]

Should we store the client URL and measure path as a global config variable(s)?

Comment (tmfrnz) [reply app/views/user_measure_mailer/created.en.html.erb:8]: Regardless, will provide you with the final text shortly - also can easily update myself.

Comment (parndt) [reply app/views/user_measure_mailer/created.en.html.erb:8]: @tmfrnz yeah, I think storing it in config is a good idea here 😄

**Comment by parndt** [comment]
>to be confirmed: activity does not have a "measure_categories" relationship with category_id = 24 ("completed", should be configurable)

What do you think of this point? Is this something I need to pay attention to?

File: app/views/user_measure_mailer/created.en.html.erb (hunk 1 above)
Comment (parndt) [reply app/views/user_measure_mailer/created.en.html.erb]: @tmfrnz I've updated the templates, let me know what you think

**Comment by tmfrnz** [comment]
> What do you think of this point? Is this something I need to pay attention to?

It should not be crucial for this type of reminder and could thus also be added later. Note, that the same condition is required for https://github.com/dumparkltd/advocacy-tracker-server/issues/48

File: app/controllers/measures_controller.rb [hunk 2]
```diff
@@ -34,7 +34,16 @@ def update
     if params[:measure][:updated_at] && DateTime.parse(params[:measure][:updated_at]).to_i != @measure.updated_at.to_i
       return render json: '{"error":"Record outdated"}', status: :unprocessable_entity
     end
+    originally_draft = @measure.draft?
```
Comment (parndt) [inline app/controllers/measures_controller.rb:37]:
```suggestion

    originally_draft = @measure.draft?
```

**Review by tmfrnz — APPROVED** [review]

File: app/controllers/user_measures_controller.rb [hunk 3]
```diff
@@ -21,6 +21,8 @@ def create
     authorize @user_measure
 
     if @user_measure.save
+      UserMeasureMailer.created(@user_measure).deliver_later if @user_measure.notify?
```
[the remark this answers is not in the document]
Comment (parndt) [reply app/controllers/user_measures_controller.rb]: yeah, it turns out that we don't yet support `deliver_later` anyway - it's an asyncronous email processing thing which uses a delayed job framework to send the email outside of the web request (it is usually used to speed up the web request)

**Comment by parndt** [comment]
@tmfrnz feel free to merge this whenever you like. I wasn't sure what you wanted to do with the email copy but that should be something you can easily update if you want or just let me know 😄

**Comment by tmfrnz** [comment]
Branch deployed to test site (also ran db:migrate), however getting a "NoMethodError" error after creating new measures (so the measures do get created but the response indicates an error)

> 2022-09-06T07:24:27.690173+00:00 app[web.1]: F, [2022-09-06T07:24:27.690092 #4] FATAL -- : [6aece5ab-a188-46e2-a030-78a8d90a9879]   
> 2022-09-06T07:24:27.690177+00:00 app[web.1]: [6aece5ab-a188-46e2-a030-78a8d90a9879] **NoMethodError (undefined method `notifications'** for #<Measure id: 105, title: "new task", description: "", target_date: nil, created_at: "2022-09-06 09:24:27.602335909 +0200", updated_at: "2022-09-06 09:24:27.602335909 +0200", draft: false, outcome: nil, indicator_summary: nil, target_date_comment: nil, updated_by_id: 1, created_by_id: 1, measuretype_id: 5, parent_id: nil, code: "", comment: "", url: "", date_start: nil, date_end: nil, date_comment: "", target_comment: nil, status_comment: nil, reference_ml: nil, reference_landbased_ml: nil, has_reference_landbased_ml: nil, status_lbs_protocol: nil, amount: nil, amount_comment: nil, private: false, is_archive: false>):
> 2022-09-06T07:24:27.690177+00:00 app[web.1]: [6aece5ab-a188-46e2-a030-78a8d90a9879]   
> 2022-09-06T07:24:27.690178+00:00 app[web.1]: [6aece5ab-a188-46e2-a030-78a8d90a9879] app/controllers/application_controller.rb:34:in `serialize'
> 2022-09-06T07:24:27.690178+00:00 app[web.1]: [6aece5ab-a188-46e2-a030-78a8d90a9879] app/controllers/measures_controller.rb:83:in `serialize'
> 2022-09-06T07:24:27.690179+00:00 app[web.1]: [6aece5ab-a188-46e2-a030-78a8d90a9879] app/controllers/measures_controller.rb:26:in `create'

**Comment by tmfrnz** [comment]
Happy to confirm that the two types of email notifications are working well otherwise!

So it would be great to hear your thoughts @parndt re. the NoMethod error. Thanks

**Comment by tmfrnz** [comment]
> @tmfrnz feel free to merge this whenever you like. I wasn't sure what you wanted to do with the email copy but that should be something you can easily update if you want or just let me know 😄

and yes, happy to adjust the email copy myself

**Comment by parndt** [comment]
@tmfrnz I noticed the link wasn't properly formed in the email from the app logs so I fixed that 🤦

**Comment by tmfrnz** [comment]
> Happy to confirm that the two types of email notifications are working well otherwise!
> 
> So it would be great to hear your thoughts @parndt re. the NoMethod error. Thanks

@parndt this no longer appears to be an issue. thanks

**Comment by tmfrnz** [comment]
@parndt sorry I busted the spec with my recent updates - will try to fix this tomorrow and let you know should I need any assistance

**Comment by parndt** [comment]
@tmfrnz I fixed the specs for you 😄

**Comment by tmfrnz** [comment]
@parndt hmm maybe the merge broke it again?

> Failures:
>
> 1) UserMeasureMailer created mentions the measure title
>   Failure/Error: expect(mail.body.encoded).to match(user_measure.measure.title)

**Pull request merged**